Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp

```diff
@@ -73,42 +73,57 @@ static GUniquePtr<char> cursorsPath(const char* basePath, Vector<GUniquePtr<char
+        return nullptr;
+
+    Vector<GUniquePtr<char>> inheritedThemes;
+    auto path = cursorsPath(basePath, inheritedThemes);
+    if (!path && inheritedThemes.isEmpty())
+        return nullptr;
+
+    if (!path) {
+        // If there's no cursors path, use the first inherited theme.
+        path = WTFMove(inheritedThemes[0]);
+        inheritedThemes.removeAt(0);
+    }
+
+    return makeUnique<CursorTheme>(WTFMove(path), size, WTFMove(inheritedThemes));
+}
+
 std::unique_ptr<CursorTheme> CursorTheme::create(const char* name, uint32_t size)
 {
-    auto findThemePath = [](const char* name) -> GUniquePtr<char> {
-        GUniquePtr<char> path(g_build_filename(g_get_user_data_dir(), "icons", name, nullptr));
-        if (g_file_test(path.get(), G_FILE_TEST_IS_DIR))
-            return path;
-
-        path.reset(g_build_filename(g_get_home_dir(), ".icons", name, nullptr));
-        if (g_file_test(path.get(), G_FILE_TEST_IS_DIR))
-            return path;
-
-        auto* dataDirs = g_get_system_data_dirs();
-        for (unsigned i = 0; dataDirs[i]; ++i) {
-            path.reset(g_build_filename(dataDirs[i], "icons", name, nullptr));
-            if (g_file_test(path.get(), G_FILE_TEST_IS_DIR))
-                return path;
-        }
+    GUniquePtr<char> path(g_build_filename(g_get_user_data_dir(), "icons", name, nullptr));
+    if (auto theme = tryCreateTheme(path.get(), size))
+        return theme;
+
+    path.reset(g_build_filename(g_get_home_dir(), ".icons", name, nullptr));
+    if (auto theme = tryCreateTheme(path.get(), size))
+        return theme;
+
+    auto* dataDirs = g_get_system_data_dirs();
+    for (unsigned i = 0; dataDirs[i]; ++i) {
+        path.reset(g_build_filename(dataDirs[i], "icons", name, nullptr));
+        if (auto theme = tryCreateTheme(path.get(), size))
+            return theme;
+    }
+
+    // If not the default, try with the default.
+    if (g_strcmp0(name, "default")) {
+        if (auto theme = CursorTheme::create("default", size))
+            return theme;
+
+        // The Adwaita fallback has already been tried.
         return nullptr;
-    };
+    }
 
-    if (auto basePath = findThemePath(name)) {
-        Vector<GUniquePtr<char>> inheritedThemes;
-        auto path = cursorsPath(basePath.get(), inheritedThemes);
-        if (path || !inheritedThemes.isEmpty()) {
-            if (!path) {
-                // If there's no cursors path, use the first inherited theme.
-                path = WTFMove(inheritedThemes[0]);
-                inheritedThemes.removeAt(0);
-            }
-
-            return makeUnique<CursorTheme>(WTFMove(path), size, WTFMove(inheritedThemes));
-        }
-        g_warning("Could not find any cursors (search started from '%s')", basePath.get());
-    } else
-        g_warning("Could not find any base paths for cursors search");
+    // Fallback to Adwaita.
+    if (auto theme = tryCreateTheme("/usr/share/icons/Adwaita/", size))
```

Comment:
The `Adwaita` theme could be installed elsewhere in any of the paths returned by `g_get_system_data_dirs()`, so I think it would make more sense to try here with `CursorTheme::create()` in the same way as done for the `default` one, instead of hardcoding the `/usr/share/icons/Adwaita` path:

```suggestion
    if (auto theme = CursorTheme::create("Adwaita", size))
```

Reply:
Good point, I'll rewrite to iterate a loop of theme names, avoiding also the recursive calls when falling back.